pkg/storage/azure/azure.go

```diff
@@ -454,162 +453,228 @@ func (d *driver) StorageChanged(cr *imageregistryv1.Config) bool {
+			cr,
+			defaults.StorageExists,
+			operatorapiv1.ConditionFalse,
+			storageExistsReasonNotConfigured,
+			"Storage account key is provided, but account name is not specified",
+		)
+		return
+	}
+
+	if d.Config.Container == "" {
+		util.UpdateCondition(
+			cr,
+			defaults.StorageExists,
+			operatorapiv1.ConditionFalse,
+			storageExistsReasonNotConfigured,
+			"Storage account is provided, but container is not specified",
+		)
+		return
+	}
+
+	cr.Status.StorageManaged = false
+	cr.Status.Storage = imageregistryv1.ImageRegistryConfigStorage{
+		Azure: d.Config.DeepCopy(),
+	}
 
-			break
+	util.UpdateCondition(
+		cr,
+		defaults.StorageExists,
+		operatorapiv1.ConditionTrue,
+		storageExistsReasonUserManaged,
+		"Storage is managed by the user",
+	)
+}
+
+// CreateStorage attempts to create a storage account and a storage container.
+func (d *driver) CreateStorage(cr *imageregistryv1.Config) error {
+	cfg, err := GetConfig(d.Listers.Secrets)
+	if err != nil {
+		util.UpdateCondition(
+			cr,
+			defaults.StorageExists,
+			operatorapiv1.ConditionUnknown,
+			storageExistsReasonConfigError,
+			fmt.Sprintf("Unable to get configuration: %s", err),
+		)
+		return err
+	}
+
+	// if AccountKey is present in our configuration it means it was provided by the user
+	// so we only verify if everything we need is in place.
+	if cfg.AccountKey != "" {
+		d.verifyUPIConfig(cr)
+		return nil
+	}
+
+	infra, err := util.GetInfrastructure(d.Listers)
+	if err != nil {
+		util.UpdateCondition(
+			cr,
+			defaults.StorageExists,
+			operatorapiv1.ConditionUnknown,
+			storageExistsReasonConfigError,
+			fmt.Sprintf("Unable to get infrastructure: %s", err),
+		)
+		return err
+	}
+
+	if d.Config.CloudName == "" && d.Config.AccountName == "" {
+		platformStatus := infra.Status.PlatformStatus
+		if platformStatus != nil &&
+			platformStatus.Type == configv1.AzurePlatformType &&
+			platformStatus.Azure != nil {
+			d.Config.CloudName = string(platformStatus.Azure.CloudName)
 		}
 	}
+
+	storageAccountName, storageAccountCreated, err := d.assureStorageAccount(cfg, infra)
+	if err != nil {
+		util.UpdateCondition(
+			cr,
+			defaults.StorageExists,
+			operatorapiv1.ConditionUnknown,
+			storageExistsReasonAzureError,
+			fmt.Sprintf("Unable to process storage account: %s", err),
+		)
+		return err
+	}
+	d.Config.AccountName = storageAccountName
+
+	containerName, containerCreated, err := d.assureContainer(cfg)
+	if err != nil {
+		util.UpdateCondition(
+			cr,
+			defaults.StorageExists,
+			operatorapiv1.ConditionUnknown,
+			storageExistsReasonAzureError,
+			fmt.Sprintf("Unable to process storage container: %s", err),
+		)
+		return err
+	}
+	d.Config.Container = containerName
+
+	cr.Spec.Storage.Azure = d.Config.DeepCopy()
+	cr.Status.Storage = imageregistryv1.ImageRegistryConfigStorage{
+		Azure: d.Config.DeepCopy(),
+	}
+
+	if storageAccountCreated && containerCreated {
+		cr.Status.StorageManaged = true
+	} else {
+		cr.Status.StorageManaged = false
```

Comment:
afaik the previous version didn't do this, and this might be wrong
if someone decided to change the container name and keep autogenerated account, the storage should still be managed by the operator (because everything is created by the operator)